### ISSUE TITLE
DataStore not saving new Users with current userId

### ISSUE THREAD

**Issue by harlanerskine** [aws-amplify/amplify-flutter#2169]
### Description

New users are not saved to Amplify. They are saved to the device but are not synced to Amplify.

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [X] DataStore
- [ ] Storage

### Steps to Reproduce

1. A new user logs in
2. On create_account_Screen they save their user info with:

```
form.save();
              developer.log(
                  'profileRepro.profilepic after form.save ${profileRepo.profilePic}');

              profileRepo.saveUserProfileDetails().then(
                (_) {
                  developer.log('saved to database');
                  developer.log(
                      'profileRepo.userId after profileRepo.saveUserProfileDetails: ${profileRepo.userId}');
                  Navigator.pushReplacement(
                    context,
                    MaterialPageRoute(
                      builder: (context) {
                        return MultiProvider(
                          providers: [
                            ChangeNotifierProvider(
                                create: (_) => ProfileRepository.instance()),
                            ChangeNotifierProvider(
                              create: (_) => SharedPrefsUtils.instance(),
                            ),
                          ],
                          child: const Home(
                              // userId: profileRepo.userId
                              ),
```

**saveUserProfileDetails in my profileRepo is:**

```
Future<void> saveUserProfileDetails() async {
    loading = true;
    try {
      developer.log('userId at save: $userId');

      final newUser = User(
        id: userId,
        username: username,
        firstName: firstNameController.text.trim(),
        lastName: lastNameController.text.trim(),
        isVerified: true,
        profilePicUrl: profilePic,
        email: email,
        userTitle: userTitleController.text.trim(),
        userLocation: userLocationController.text.trim(),
        userBio: bioController.text,
        userBioUrl: bioUrlController.text,
        createdOn: TemporalDateTime.now(),
      );

      await Amplify.DataStore.save(newUser).then((_) => loading = false);
    } catch (e) {
      developer.log('Could not save DataStore: $e');
    }
  }
```

userId at save: $userId logs the current user that should be connected to that info.
but no save is sent to Amplify.

If I remove the line:
`id: userId`
a user is saved.

my schema.graphql is:
```
type CopyrightApplication @model @auth(rules: [{allow: public}]) @key(name: "byUser", fields: ["userID"]) {
  id: ID!
  Posts: [Post] @connection(keyName: "byCopyrightApplication", fields: ["id"])
  applicationStatus: CopyrightStatus!
  applicationName: String!
  userID: ID!
  applicationCreatedOn: AWSDateTime!
  applicationSubmittedOn: AWSDateTime
}

type Timeline @model @auth(rules: [{allow: public}]) {
  id: ID!
  Posts: [Post] @connection(keyName: "byTimeline", fields: ["id"])
  Followings: [Following] @connection(keyName: "byTimeline", fields: ["id"])
}

type Followers @model @auth(rules: [{allow: public}]) @key(name: "byUser", fields: ["userID"]) {
  id: ID!
  userID: ID!
}

type Following @model @auth(rules: [{allow: public}]) @key(name: "byUser", fields: ["userID"]) @key(name: "byTimeline", fields: ["timelineID"]) {
  id: ID!
  userID: ID!
  timelineID: ID!
}

enum PostStatus {
  CREATED
  UPDATED
  DELETED
}

enum CopyrightStatus {
  POSTED
  INQUEUE
  APPLIED
  ERROR
  REGISTERED
}

type Post @model @key(name: "byUser", fields: ["userID"]) @auth(rules: [{allow: public}, {allow: owner}]) @key(name: "byTimeline", fields: ["timelineID"]) @key(name: "byCopyrightApplication", fields: ["copyrightapplicationID"]) {
  id: ID!
  caption: String!
  postImageUrl: String!
  copyrightStatus: CopyrightStatus
  location: String
  createdOn: AWSDateTime!
  updatedOn: AWSDateTime
  postStatus: PostStatus!
  user: User @connection
  comments: [Comment] @connection(keyName: "byPost", fields: ["id"])
  likes: String
  userID: ID!
  timelineID: ID!
  copyrightapplicationID: ID!
}

type Comment @model @key(name: "byPost", fields: ["postID"]) @auth(rules: [{allow: public}, {allow: owner}]) {
  id: ID!
  createdOn: AWSDateTime!
  updatedOn: AWSDateTime
  userId: String!
  commentText: String!
  post: Post @connection
  postID: ID!
}

type User @model @auth(rules: [{allow: public}]) {
  id: ID!
  username: String!
  firstName: String!
  lastName: String!
  isVerified: Boolean!
  profilePicUrl: String
  email: AWSEmail!
  userTitle: String
  userLocation: String
  posts: [Post!] @connection(keyName: "byUser", fields: ["id"])
  userBio: String
  userBioUrl: String
  followers: [Followers!] @connection(keyName: "byUser", fields: ["id"])
  following: [Following!] @connection(keyName: "byUser", fields: ["id"])
  createdOn: AWSDateTime!
  updatedOn: AWSDateTime
  CopyrightApplications: [CopyrightApplication] @connection(keyName: "byUser", fields: ["id"])
}
```

### Platforms

- [X] iOS
- [ ] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Environment

```bash
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.3.2, on macOS 12.5.1 21G83 darwin-x64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] IntelliJ IDEA Community Edition (version 2020.2.3)
[✓] VS Code (version 1.71.2)
[✓] Connected device (4 available)
[✓] HTTP Host Availability

• No issues found!
```

### Dependencies

```bash
Dart SDK 2.18.1
Flutter SDK 3.3.2
super_lux 0.1.0+16

dependencies:
- amplify_analytics_pinpoint 0.6.7 [amplify_analytics_pinpoint_android amplify_analytics_pinpoint_ios amplify_core aws_common flutter meta]
- amplify_api 0.6.7 [amplify_api_android amplify_api_ios amplify_core amplify_flutter aws_common collection flutter meta plugin_platform_interface]
- amplify_auth_cognito 0.6.7 [amplify_auth_cognito_android amplify_auth_cognito_ios amplify_core aws_common collection flutter meta plugin_platform_interface]
- amplify_authenticator 0.2.2 [amplify_auth_cognito amplify_core amplify_flutter aws_common collection flutter flutter_localizations intl]
- amplify_datastore 0.6.7 [flutter amplify_datastore_plugin_interface amplify_core plugin_platform_interface meta collection async]
- amplify_flutter 0.6.7 [amplify_core amplify_datastore_plugin_interface amplify_flutter_android amplify_flutter_ios aws_common collection flutter meta plugin_platform_interface]
- amplify_storage_s3 0.6.7 [amplify_storage_s3_android amplify_storage_s3_ios amplify_core aws_common flutter meta plugin_platform_interface]
- animator 3.2.0 [flutter states_rebuilder]
- cached_network_image 3.2.2 [flutter flutter_cache_manager octo_image cached_network_image_platform_interface cached_network_image_web]
- cupertino_icons 1.0.5
- cupertino_list_tile 0.2.1 [cupertino_icons flutter]
- email_validator 2.1.17
- flutter 0.0.0 [characters collection material_color_utilities meta vector_math sky_engine]
- font_awesome_flutter 10.1.0 [flutter]
- get 4.6.5 [flutter]
- image_cropper 3.0.0 [flutter image_cropper_platform_interface image_cropper_for_web]
- image_picker 0.8.5+3 [flutter image_picker_android image_picker_for_web image_picker_ios image_picker_platform_interface]
- path_provider 2.0.11 [flutter path_provider_android path_provider_ios path_provider_linux path_provider_macos path_provider_platform_interface path_provider_windows]
- provider 6.0.3 [collection flutter nested]
- pull_to_refresh_flutter3 2.0.1 [flutter]
- shared_preferences 2.0.15 [flutter shared_preferences_android shared_preferences_ios shared_preferences_linux shared_preferences_macos shared_preferences_platform_interface shared_preferences_web shared_preferences_windows]
- timeago 3.3.0 [intl]
- url_launcher 6.1.5 [flutter url_launcher_android url_launcher_ios url_launcher_linux url_launcher_macos url_launcher_platform_interface url_launcher_web url_launcher_windows]
- uuid 3.0.6 [crypto]

transitive dependencies:
- amplify_analytics_pinpoint_android 0.6.7 [flutter]
- amplify_analytics_pinpoint_ios 0.6.7 [flutter]
- amplify_api_android 0.6.7 [flutter]
- amplify_api_ios 0.6.7 [amplify_core flutter]
- amplify_auth_cognito_android 0.6.7 [flutter]
- amplify_auth_cognito_ios 0.6.7 [amplify_core flutter]
- amplify_core 0.6.7 [aws_common collection flutter intl json_annotation meta plugin_platform_interface uuid]
- amplify_datastore_plugin_interface 0.6.7 [amplify_core collection flutter meta]
- amplify_flutter_android 0.6.7 [flutter]
- amplify_flutter_ios 0.6.7 [amplify_core flutter]
- amplify_storage_s3_android 0.6.7 [flutter]
- amplify_storage_s3_ios 0.6.7 [flutter]
- async 2.9.0 [collection meta]
- aws_common 0.1.1 [async collection http meta stream_transform uuid]
- cached_network_image_platform_interface 2.0.0 [flutter flutter_cache_manager]
- cached_network_image_web 1.0.2 [flutter flutter_cache_manager cached_network_image_platform_interface]
- characters 1.2.1
- clock 1.1.1
- collection 1.16.0
- cross_file 0.3.3+1 [js meta]
- crypto 3.0.2 [typed_data]
- ffi 2.0.1
- file 6.1.2 [meta path]
- flutter_blurhash 0.7.0 [flutter]
- flutter_cache_manager 3.3.0 [clock collection file flutter http path path_provider pedantic rxdart sqflite uuid]
- flutter_localizations 0.0.0 [flutter intl characters clock collection material_color_utilities meta path vector_math]
- flutter_plugin_android_lifecycle 2.0.7 [flutter]
- flutter_web_plugins 0.0.0 [flutter js characters collection material_color_utilities meta vector_math]
- http 0.13.5 [async http_parser meta path]
- http_parser 4.0.1 [collection source_span string_scanner typed_data]
- image_cropper_for_web 1.0.2 [flutter flutter_web_plugins image_cropper_platform_interface js]
- image_cropper_platform_interface 3.0.2 [flutter plugin_platform_interface http]
- image_picker_android 0.8.5+2 [flutter flutter_plugin_android_lifecycle image_picker_platform_interface]
- image_picker_for_web 2.1.8 [flutter flutter_web_plugins image_picker_platform_interface]
- image_picker_ios 0.8.5+6 [flutter image_picker_platform_interface]
- image_picker_platform_interface 2.6.1 [cross_file flutter http plugin_platform_interface]
- intl 0.17.0 [clock path]
- js 0.6.4
- json_annotation 4.6.0 [meta]
- material_color_utilities 0.1.5
- meta 1.8.0
- nested 1.0.0 [flutter]
- octo_image 1.0.2 [flutter flutter_blurhash]
- path 1.8.2
- path_provider_android 2.0.19 [flutter path_provider_platform_interface]
- path_provider_ios 2.0.11 [flutter path_provider_platform_interface]
- path_provider_linux 2.1.7 [ffi flutter path path_provider_platform_interface xdg_directories]
- path_provider_macos 2.0.6 [flutter path_provider_platform_interface]
- path_provider_platform_interface 2.0.4 [flutter platform plugin_platform_interface]
- path_provider_windows 2.1.2 [ffi flutter path path_provider_platform_interface win32]
- pedantic 1.11.1
- platform 3.1.0
- plugin_platform_interface 2.1.2 [meta]
- process 4.2.4 [file path platform]
- rxdart 0.27.5
- shared_preferences_android 2.0.12 [flutter shared_preferences_platform_interface]
- shared_preferences_ios 2.1.1 [flutter shared_preferences_platform_interface]
- shared_preferences_linux 2.1.1 [file flutter path path_provider_linux path_provider_platform_interface shared_preferences_platform_interface]
- shared_preferences_macos 2.0.4 [flutter shared_preferences_platform_interface]
- shared_preferences_platform_interface 2.0.0 [flutter]
- shared_preferences_web 2.0.4 [flutter flutter_web_plugins shared_preferences_platform_interface]
- shared_preferences_windows 2.1.1 [file flutter path path_provider_platform_interface path_provider_windows shared_preferences_platform_interface]
- sky_engine 0.0.99
- source_span 1.9.0 [collection path term_glyph]
- sqflite 2.0.3+1 [flutter sqflite_common path]
- sqflite_common 2.2.1+1 [synchronized path meta]
- states_rebuilder 6.1.0+1 [flutter flutter_localizations collection meta]
- stream_transform 2.0.0
- string_scanner 1.1.1 [source_span]
- synchronized 3.0.0+2
- term_glyph 1.2.1
- typed_data 1.3.1 [collection]
- url_launcher_android 6.0.17 [flutter url_launcher_platform_interface]
- url_launcher_ios 6.0.17 [flutter url_launcher_platform_interface]
- url_launcher_linux 3.0.1 [flutter url_launcher_platform_interface]
- url_launcher_macos 3.0.1 [flutter url_launcher_platform_interface]
- url_launcher_platform_interface 2.1.0 [flutter plugin_platform_interface]
- url_launcher_web 2.0.13 [flutter flutter_web_plugins url_launcher_platform_interface]
- url_launcher_windows 3.0.1 [flutter url_launcher_platform_interface]
- vector_math 2.1.2
- win32 2.7.0 [ffi]
- xdg_directories 0.2.0+1 [meta path process]
```

### Device

iPhone 12 Pro, iPhone 14 Pro

### OS

iOS 16.0 and below

### Deployment Method

Amplify CLI

### CLI Version

10.0.0

### Amplify Config

const amplifyconfig = ''' {
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "analytics": {
        "plugins": {
            "awsPinpointAnalyticsPlugin": {
                "pinpointAnalytics": {
                    "appId": "...",
                    "region": "us-east-1"
                },
                "pinpointTargeting": {
                    "region": "us-east-1"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify-cli/0.1.0",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "us-east-...",
                            "Region": "us-east-1"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "us-east-1_5ijJr5CgH",
                        "AppClientId": "...",
                        "Region": "us-east-1"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH",
                        "socialProviders": [],
                        "usernameAttributes": [],
                        "signupAttributes": [
                            "EMAIL"
                        ],
                        "passwordProtectionSettings": {
                            "passwordPolicyMinLength": 8,
                            "passwordPolicyCharacters": []
                        },
                        "mfaConfiguration": "OPTIONAL",
                        "mfaTypes": [
                            "SMS"
                        ],
                        "verificationMechanisms": [
                            "EMAIL"
                        ]
                    }
                },
                "PinpointAnalytics": {
                    "Default": {
                        "AppId": "...",
                        "Region": "us-east-1"
                    }
                },
                "PinpointTargeting": {
                    "Default": {
                        "Region": "us-east-1"
                    }
                },
                "AppSync": {
                    "Default": {
                        "ApiUrl": "...",
                        "Region": "us-east-1",
                        "AuthMode": "API_KEY",
                        "ApiKey": "da2-...",
                        "ClientDatabasePrefix": "..."
                    },
                    "superlux_AWS_IAM": {
                        "ApiUrl": "...",
                        "Region": "us-east-1",
                        "AuthMode": "AWS_IAM",
                        "ClientDatabasePrefix": "superlux_AWS_IAM"
                    },
                    "superlux_AMAZON_COGNITO_USER_POOLS": {
                        "ApiUrl": "...",
                        "Region": "us-east-1",
                        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
                        "ClientDatabasePrefix": "superlux_AMAZON_COGNITO_USER_POOLS"
                    }
                },
                "S3TransferUtility": {
                    "Default": {
                        "Bucket": "...-dev",
                        "Region": "us-east-1"
                    }
                }
            }
        }
    },
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "superlux": {
                    "endpointType": "GraphQL",
                    "endpoint": "...",
                    "region": "us-east-1",
                    "authorizationType": "API_KEY",
                    "apiKey": "da2-..."
                }
            }
        }
    },
    "storage": {
        "plugins": {
            "awsS3StoragePlugin": {
                "bucket": "...",
                "region": "us-east-1",
                "defaultAccessLevel": "guest"
            }
        }
    }
}''';

**Comment by fjnoyp** [comment]
Hi @harlanerskine where do you get `userId` ?

How do you know it is saved locally but not synced to AWS - is there an error you get in the logs that you can share?

**Comment by harlanerskine** [comment]
Hi @fjnoyp, thanks for looking this issue over. 

The `userId` is logged as the correct id in 
`developer.log('userId at save: $userId');`
at the beginning of the try block in my save

`Future<void> saveUserProfileDetails() async {`

userId at save: 55ebfe43-c690-420c...

I think it's not synced up to AWS because, in my login, I have the code:

`onPressed: () {
                                    if (_formLoginKey.currentState!.validate()) {
                                      loginRep
                                          .login(context)
                                          .then((List<User>? userlist) {
                                        if (userlist == null) {
                                        } else {
                                          developer.log(userlist.toString());
                                          if (userlist.isNotEmpty) {
                                            Navigator.of(context).push(
                                                CupertinoPageRoute(
                                                    builder: (context) {
                                              // Navigator.push(context,
                                              //     MaterialPageRoute(
                                              //         builder: (context) {
                                              return ChangeNotifierProvider(
                                                create: (_) => ProfileRepository
                                                    .instance(),
                                                child: const Home(),
                                              );
                                            }));
                                          } else {
                                            Navigator.of(context).push(
                                                CupertinoPageRoute(
                                                    builder: (context) {
                                              // Navigator.push(context,
                                              //     MaterialPageRoute(
                                              //         builder: (context) {
                                              return ChangeNotifierProvider(
                                                create: (_) => ProfileRepository
                                                    .instance(),
                                                child:
                                                    const CreateAccountScreen(),
                                              );
                                            }));
                                          }
                                        }`
And if the db is clear, the CreateAccountScreen() is displayed, but if that screen has been saved and the user logged out, the Home() screen is displayed on login.

Thanks for your input on this.

**Comment by Jordan-Nelson** [comment]
Hi @harlanerskine - Can you run the app from Xcode and see if there are any error messages in the logs? Amplify Flutter currently depends on Amplify-iOS when targeting iOS, and the logs from Amplify-iOS will not show up when running it from VS Code or Android Studio. If you use VS Code with the flutter plugin, you can right click on the /ios dir and choose "Open in Xcode".

**Comment by harlanerskine** [comment]
@Jordan-Nelson Thanks. I'll try that and get back to you. Would I need to run the simulator from Xcode? I'm having trouble getting things to run (which is a separate issue). I'll comment back here when there are logs to report.

**Comment by Jordan-Nelson** [comment]
@harlanerskine as long as you have a simulator running, it shouldn't matter where it was opened from.

**Comment by harlanerskine** [comment]
Ok - I got the simulator back up (I needed to delete one of the models that was causing a problem (separate from this error). I ran the app in Xcode and I have the following log:

`Encoder Foundation.(unknown context at $7ff800dd0908).__JSONEncoder
Successfully saved model: FlutterSerializedModel(values: ["createdOn": Amplify.JSONValue.string("2022-11-10T21:38:40.965777000Z"), "createdAt": Amplify.JSONValue.null, "updatedAt": Amplify.JSONValue.null, "userLocation": Amplify.JSONValue.string("id: userId"), "userBio": Amplify.JSONValue.string("bio test"), "id": Amplify.JSONValue.string("55ebfe43-c690-420c-990c-05d94802a07e"), "username": Amplify.JSONValue.string("harlanerskine"), "userTitle": Amplify.JSONValue.string("4:38pm"), "userEmail": Amplify.JSONValue.string("harlanerskine@gmail.com"), "isVerified": Amplify.JSONValue.boolean(true), "lastName": Amplify.JSONValue.string("Thursday"), "__typename": Amplify.JSONValue.string("User"), "userBioUrl": Amplify.JSONValue.string("www.test.com"), "profilePicUrl": Amplify.JSONValue.string(""), "updatedOn": Amplify.JSONValue.null, "firstName": Amplify.JSONValue.string("testing"), "ownerId": Amplify.JSONValue.string("temp")], _modelName: "User")
Failed to parse the event DataStoreError: No decoder found for model named MutationEvent
Recovery suggestion: There is no decoder registered for the model named MutationEvent. Register models with `ModelRegistry.register(modelName:)` at startup.
Encoder Foundation.(unknown context at $7ff800dd0908).__JSONEncoder`
